specifyweb/frontend/js_src/lib/components/Statistics/index.tsx

```diff
@@ -0,0 +1,805 @@
+import { f } from '../../utils/functools';
+import { removeItem, removeKey, replaceItem } from '../../utils/utils';
+import { H2, H3 } from '../Atoms';
+import { Button } from '../Atoms/Button';
+import { className } from '../Atoms/className';
+import { Form } from '../Atoms/Form';
+import { Submit } from '../Atoms/Submit';
+import { softFail } from '../Errors/Crash';
+import { useMenuItem } from '../Header';
+import { userInformation } from '../InitialContext/userInformation';
+import { DateElement } from '../Molecules/DateElement';
+import { downloadFile } from '../Molecules/FilePicker';
+import { hasPermission } from '../Permissions/helpers';
+import { useQueries } from '../Toolbar/Query';
+import { awaitPrefsSynced } from '../UserPreferences/helpers';
+import { useCollectionPref, usePref } from '../UserPreferences/usePref';
+import { AddStatDialog } from './AddStatDialog';
+import { StatsPageButton } from './Buttons';
+import { Categories } from './Categories';
+import {
+  statsToTsv,
+  useBackendApi,
+  useCategoryToFetch,
+  useDefaultLayout,
+  useDefaultStatsToAdd,
+  useStatsSpec,
+  useUnknownCategory,
+} from './hooks';
+import { StatsPageEditing } from './StatsPageEditing';
+import type { CustomStat, DefaultStat, StatLayout } from './types';
+import { urlSpec } from './definitions';
+import { RA } from '../../utils/types';
+import { cleanFulfilledRequests } from '../../utils/ajax/throttledAjax';
+import { useTriggerState } from '../../hooks/useTriggerState';
+import { getUniqueName } from '../../utils/uniquifyName';
+
+export function StatsPage(): JSX.Element | null {
+  useMenuItem('statistics');
+  const [collectionLayout, setCollectionLayout] = useCollectionPref(
+    'statistics',
+    'appearance',
+    'layout'
+  );
+
+  const [personalLayout, setPersonalLayout] = usePref(
+    'statistics',
+    'appearance',
+    'layout'
+  );
+
+  const [defaultLayout, setDefaultLayout] = useCollectionPref(
+    'statistics',
+    'appearance',
+    'defaultLayout'
+  );
+
+  const layout = {
+    [statsText.shared()]: collectionLayout,
+    [statsText.personal()]: personalLayout,
+  };
+
+  const [state, setState] = React.useState<
+    | State<
+        'AddingState',
+        {
+          readonly pageIndex: number;
+          readonly categoryIndex: number;
+        }
+      >
+    | State<
+        'PageRenameState',
+        {
+          readonly pageIndex: number | undefined;
+          readonly isCollection: boolean;
+        }
+      >
+    | State<'DefaultState'>
+    | State<'EditingState'>
+    | State<
+        'DeletingCategoryState',
+        { readonly categoryContainsCustom: boolean }
+      >
+  >({ type: 'DefaultState' });
+
+  const isAddingItem = state.type === 'AddingState';
+  const isEditing =
+    state.type === 'EditingState' ||
+    isAddingItem ||
+    state.type === 'PageRenameState';
+  const isEditingCollection = React.useRef(false);
+  const canEditIndex = (isCollection: boolean): boolean =>
```

Comment:
In this case, it works since canEditIndex gets called only when isEditing (which is part of state) is true (see the use cases of canEditIndex), and isEditingCollection is changed only when the user goes into editing state (when isEditing is set to true) so we always have the updated value.